pacman/operations/rigged_algorithms/hilbert_placer.py

```diff
@@ -0,0 +1,171 @@
+
+
+class HilbertPlacer(object):
+
+    def __call__(self, machine_graph, machine):
+
+        # check that the algorithm can handle the constraints
+        self._check_constraints(machine_graph.vertices)
+
+        # in order to test isomorphism:
+        # placements_copy = Placements()
+        placements = Placements()
+        vertices = \
+            placer_algorithm_utilities.sort_vertices_by_known_constraints(
+                machine_graph.vertices)
+
+        # Iterate over vertices and generate placements
+        progress = ProgressBar(machine_graph.n_vertices,
+                               "Placing graph vertices")
+        resource_tracker = ResourceTracker(machine,
+                                           self._generate_hilbert_chips(
+                                               machine))
+        vertices_on_same_chip = \
+            placer_algorithm_utilities.get_same_chip_vertex_groups(
+                    machine_graph.vertices)
+        all_vertices_placed = set()
+        for vertex in progress.over(vertices):
+            if vertex not in all_vertices_placed:
+                vertices_placed = self._place_vertex(
+                    vertex, resource_tracker, machine,
+                    # placements_copy,
+                    placements,
+                    vertices_on_same_chip)
+                all_vertices_placed.update(vertices_placed)
+        # return placements_copy
+        return placements
+
+    def _check_constraints(
+            self, vertices, additional_placement_constraints=None):
+        placement_constraints = {SameChipAsConstraint}
+        if additional_placement_constraints is not None:
+            placement_constraints.update(additional_placement_constraints)
+        ResourceTracker.check_constraints(
+            vertices, additional_placement_constraints=placement_constraints)
+
+    def _hilbert_curve(self, level, angle=1, state=None):
+        """Generator of points along a 2D Hilbert curve.
+
+        This implements the L-system as described on
+        `http://en.wikipedia.org/wiki/Hilbert_curve`.
+
+        :param level: Number of levels of recursion to use in generating
+            the curve. The resulting curve will be `(2**level)-1` wide/tall.
+        :type level: int
+        :param angle: `1` if this is the 'positive'
+            expansion of the grammar and `-1` for the 'negative' expansion.
+        :type angle: int
+        :param state: The current state of the system in a Hilbert curve.
+        :type state: \
+            :py:class:`pacman.operations.rigged_algorithms.hilbert_state.py`
+        """
+
+        # Create state object first time we're called while also yielding first
+        # position
+        if state is None:
+            state = HilbertState()
+            yield state.x_pos, state.y_pos
+
+        if level <= 0:
+            return
+
+        # Turn left
+        state.change_x, state.change_y = state.change_y * -angle, \
```

Comment:
nah, and if so, you could split the lines into 2, which is just as easy to read. You gain no performance increase doing both in 1 line. 

state.change_x =  state.change_y * -angle
state,change_y = .... whatever this was